pyface/data_view/value_types/enum_value.py

```diff
@@ -0,0 +1,162 @@
+        row : sequence of int
+            The row in the data model being queried.
+        column : sequence of int
+            The column in the data model being queried.
+        value : any
+            The value to validate.
+
+        Returns
+        -------
+        is_valid : bool
+            Whether or not the value is valid.
+        """
+        return value in self.values
+
+    def has_text(self, model, row, column):
+        """ Get the display text from the underlying value.
+
+        Parameters
+        ----------
+        model : AbstractDataModel
+            The data model holding the data.
+        row : sequence of int
+            The row in the data model being queried.
+        column : sequence of int
+            The column in the data model being queried.
+
+        Returns
+        -------
+        text : str
+            The text to display.
+        """
+        return self.format is not None
+
+    def get_text(self, model, row, column):
+        """ Get the display text from the underlying value.
+
+        Parameters
+        ----------
+        model : AbstractDataModel
+            The data model holding the data.
+        row : sequence of int
+            The row in the data model being queried.
+        column : sequence of int
+            The column in the data model being queried.
+
+        Returns
+        -------
+        text : str
+            The text to display.
+        """
+        return self.format(model.get_value(row, column))
+
+    def has_color(self, model, row, column):
+        """ Whether or not the value has color data.
+
+        has_color : bool
+            Whether or not the value has data-associated color
+            values.
+        """
+        return (
+            self.colors is not None
+            and self.colors(model.get_value(row, column)) is not None
+        )
+
+    def get_color(self, model, row, column):
+        """ Get data-associated colour values for the given item.
+
+        The default implementation returns white.
+
+        color : Color instance
+            The color associated with the cell.
+        """
+        return self.colors(model.get_value(row, column))
+
+    def has_image(self, model, row, column):
+        """ Whether or not the value has an image associated with it.
+
+        The default implementation returns True if ``get_image``
+        returns a non-None value.
+
+        Parameters
+        ----------
+        model : AbstractDataModel
+            The data model holding the data.
+        row : sequence of int
+            The row in the data model being queried.
+        column : sequence of int
+            The column in the data model being queried.
+
+        Returns
+        -------
+        has_image : bool
+            Whether or not the value has an image associated with it.
+        """
+        return (
+            self.images is not None
+            and self.images(model.get_value(row, column)) is not None
```

Comment:
Same here.

Reply:
Image is different here that when there are no images, the cell appearance is different. But on the other hand, the data view item model seems to be ignoring anything that is not an IImageResource anyway, so returning, say `1` from `get_image`, would have no consequences.

Reply:
This prompted me to look into error handling again. When the `DataViewGetError` and `DataViewSetError` were introduced, they were dealing with getting and setting values on the data models. The `images`, `colors` and `format` are dealing with getting different representations for a given value after data model says it is okay, and we might want to have a different error handling for that which can be applied across all value types?  It seems that for some representations, it is not worth crashing the application for. e.g. if `format` fails for whatever reason, does it make sense to crash? Likewise for tooltip, is error from failing to compute a tooltip worth crashing the application?

Maybe a separate issue for this...

Reply:
Currently in the implementations if any of the getters raise `DataViewGetError` it will be ignored, so an alternative approach is to have the `get_color` method do the isinstance check and raise if it isn't a `Color`.

But returning `1` from `get_image` would cause problems - and likely a crash with error reporting to the log, because the toolkit code expecting an `IImageResource` will fail.  This is going to be true for any getter that returns a type that the toolkit code can't handle (eg. a number from `get_text` or `get_tooltip`).  This is fine based on the design we have, as these are likely developer errors so we want to fail fast with good diagnostics.

But yes, this is probably a discussion for a separate issue.

Reply:
> But returning 1 from get_image would cause problems - and likely a crash with error reporting to the log, because the toolkit code expecting an IImageResource will fail. 

For reference: I tried making `get_image` return `1` for the Qt implementation and it did not crash, it just falls through as if there are no images, because it did the `isinstance` check. I did not try with Wx though.  For color, the same cannot be said...

Reply:
Images are currently ignored by Wx - they are possible, but the column-orientation of the underlying widget makes it hard.